### PR TITLE
[WIP] An aspirational document on renewing certificates using kubeadm.

### PR DESCRIPTION
**Pull request by liztio**
Commands mentioned do not currently exist.

Part of kubernetes/kubeadm#206

### CODE REVIEW

**Comment by neolit123** [comment]
@liztio looks good, thanks.
perhaps we could have used google docs for the draft to allow collab edits?

do we have an associated issue in `k/kubeadm`?

File: content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+{{% capture overview %}}
+
+This page explains how to rotate the various certificates associated with a kubeadm k8s installation.
+
+{{% /capture %}}
+
+{{% capture steps %}}
+
+As part of upgrades, `kubeadm` will check all certificates it manages for expiration.
```
Comment (fabriziopandini) [inline content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md]: Does that means are you planning to change the upgrade process as well? currently, if I remember well it replaces only api server cert older than 180 days

File: content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+In the case of credential leaks, best practices are to create a new, uncompromised cluster.
+{{</ warning >}}
+
+## Cluster CA
+
+The cluster CA is initialised with a lifetime of [10 years][cacreation]. This is unlikely to require a reset.
+
+[ca creation]: https://github.com/kubernetes/kubernetes/blob/release-1.11/staging/src/k8s.io/client-go/util/cert/cert.go#L72
+
+## Kubelet client certificates
```
Comment (fabriziopandini) [inline content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md]: @liztio might be I don't have all the info here, but why we need this if we are using Node certificate rotation via CSR auto-approved by the csrapprover controller?

File: content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md (hunk 1 above)
Comment (liztio) [reply content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md]: Oh is _that_ how it works? I was told is was "3 months from expiry," which I was able to prove wasn't the case.

File: content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md (hunk 2 above)
Comment (liztio) [reply content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md]: I need review here from a security engineer. My thinking was a compromised certificate could allow the compromise to continue if we used the same trust paths. But maybe that's not a real risk factor?

**Comment by liztio** [comment]
@neolit123 https://github.com/kubernetes/kubeadm/issues/206

File: content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+On the master: 
+
+```
+$ kubeadm phase certs renew master
+certs bootstrap created. Run `kubeadm phase certs renew kubelet --token <some token>
+```
+
+Run the provided command on all kubelet nodes. This re-bootstraps the trust from the master to the nodes.
```
Comment (timothysc) [inline content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md]: I wonder if we can automate this now via the dynamic kubelet jiggery.  

/cc @mtaufen

File: content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md (hunk 1 above)
Comment (fabriziopandini) [reply content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md]: That is my understanding of this code at least ...
https://github.com/kubernetes/kubernetes/blob/0c8fe56ea41d59b9f795949fbced9fde6ab422eb/cmd/kubeadm/app/phases/upgrade/postupgrade.go#L304

**Comment by kachkaev** [comment]
Thanks for putting this together @liztio! Do you know if there is a way to rotate the certs _after_ the kubeadm one got expired? If yes, it'd be great to have this described in the docs as may people may struggle with this ([me in particular](https://github.com/kubernetes/kubeadm/issues/581#issuecomment-410389704) 😅)

File: content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md (hunk 1 above)
Comment (xiangpengzhao) [reply content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md]: Agree with @fabriziopandini :)

File: content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+{{% capture steps %}}
+
+As part of upgrades, `kubeadm` will check all certificates it manages for expiration.
+If they are expiring within the next month, they will be replaced.
+
+Alternatively, the following commands can be used to granularly rotate individual certificates.
+
+{{< warning >}}
+
+Kubernetes CAs do not support OCSP.
```
Comment (xiangpengzhao) [inline content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md]: Might be good to write the full words of OCSP?

File: content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+[ca creation]: https://github.com/kubernetes/kubernetes/blob/release-1.11/staging/src/k8s.io/client-go/util/cert/cert.go#L72
+
+## Kubelet client certificates
+
+On the master: 
+
+```
+$ kubeadm phase certs renew master
+certs bootstrap created. Run `kubeadm phase certs renew kubelet --token <some token>
```
Comment (xiangpengzhao) [inline content/en/docs/tasks/administer-cluster/kubeadm/rotate-certs.md]: Missing a ` at the end

**Comment by zacharysarah** [comment]
@liztio 👋 Thanks for this PR! It looks like the last activity here was 18 days ago. Please feel free to reopen this PR when you're ready to address [review feedback](https://github.com/kubernetes/website/pull/9712#pullrequestreview-151246060) and drop `[WIP]` from the PR title.

**Pull request closed without merging**